File: leetcode/add_two_numbers.py
# Jacobus Burger (2025-07-10)
# Solution to Add Two Numbers on LeetCode


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Solution
# Stats:
# - 5ms runtime, beats 42.55%
# - 17.80MB memory, beats 84.93%
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        solution = ListNode()
        head = solution
        carry = 0
        while l1 or l2 or carry != 0:
            sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            carry = sum // 10

            head.next = ListNode(val=sum % 10)
            head = head.next
            l1 = (l1.next if l1 is not None else None)
            l2 = (l2.next if l2 is not None else None)
        return solution.next

File: leetcode/add_two_numbers.c
// Jacobus Burger (2025-07-24)
// Solution to Add Two Numbers on LeetCode
#include <stdio.h>
#include <stdlib.h>


// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};


// Solution
// Stats:
// - 0ms runtime, beats 100%
// - 12.74MB memory, beats 36.68%
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode solution = {0, NULL};
    struct ListNode* current = &solution;
    int carry = 0;

    while (l1 != NULL || l2 != NULL || carry != 0) {
        int a = l1 != NULL ? l1->val : 0;
        int b = l2 != NULL ? l2->val : 0;
        int sum = a + b + carry;
        carry = sum / 10;

        struct ListNode* next = (struct ListNode*) malloc(sizeof(struct ListNode));
        next->val = sum % 10;
        next->next = NULL;

        current->next = next;
        current = current->next;
        l1 = l1 != NULL ? l1->next : NULL;
        l2 = l2 != NULL ? l2->next : NULL;
    }
    return solution.next;
}
